openedx/core/djangoapps/course_groups/models.py

```diff
@@ -58,6 +59,96 @@ def create(cls, name, course_id, group_type=COHORT):
+        self.full_clean(validate_unique=False)
+
+        max_retries = 5
+        for num_attempts in range(max_retries):
+            saved_membership = None
+            try:
+                # The following 2 "transaction" lines force a fresh read, they can be removed once we're on django 1.8
+                # http://stackoverflow.com/questions/3346124/how-do-i-force-django-to-ignore-any-caches-and-reload-data
+                with transaction.commit_manually():
+                    transaction.commit()
+
+                saved_membership = CohortMembership.objects.get(
+                    user__id=self.user.id,  # pylint: disable=E1101
+                    course_id=self.course_id
+                )
+            except CohortMembership.DoesNotExist:
+                pass
+
+            if saved_membership is not None:
+                if saved_membership.course_user_group == self.course_user_group:
+                    raise ValueError("User {user_name} already present in cohort {cohort_name}".format(
+                        user_name=self.user.username,  # pylint: disable=E1101
+                        cohort_name=self.course_user_group.name
+                    ))
+                self.previous_cohort = saved_membership.course_user_group
+                self.previous_cohort_name = saved_membership.course_user_group.name
+                self.previous_cohort_id = saved_membership.course_user_group.id
+
+            try:
+                with transaction.commit_on_success():
+                    self.course_user_group.users.add(self.user)  # pylint: disable=E1101
+                    self.user.course_groups.add(self.course_user_group)
+                    if saved_membership is not None:
+                        self.previous_cohort.users.remove(self.user)
+                        updated = CohortMembership.objects.filter(
+                            id=saved_membership.id,
+                            version=saved_membership.version
+                        ).update(
+                            course_user_group=self.course_user_group,
+                            version=saved_membership.version + 1
+                        )
+                        if not updated:
+                            raise IntegrityError("value of saved_membership has changed since read")
+                    else:
+                        super(CohortMembership, self).save(*args, **kwargs)
+            except IntegrityError:
+                if num_attempts == max_retries:
+                    raise IntegrityError("Unable to save membership after {} tries, aborting.".format(max_retries))
+                num_attempts += 1
+                continue
+
+            break
+
+    course_user_group = models.ForeignKey(CourseUserGroup)
+    user = models.ForeignKey(User)
+    version = models.IntegerField(default=0)
+    course_id = CourseKeyField(max_length=255)
+
+    previous_cohort = None
+    previous_cohort_name = None
+    previous_cohort_id = None
```

Comment:
I believe by convention, these fields are placed at the top of the class, just so it's easier to understand what these values are when they're used.